specfile/prep.py

```diff
@@ -0,0 +1,303 @@
+        self._parser.add_argument("-a", type=int)
+
+
+class PatchMacro(PrepMacro):
+    """Class that represents a %patch macro."""
+
+    def _setup_parser(self) -> None:
+        """Configures internal `ArgumentParser` for the options of the macro."""
+        self._parser.add_argument("-P", type=int)
+        self._parser.add_argument("-p", type=int)
+        self._parser.add_argument("-b")
+        self._parser.add_argument("-E", action="store_true")
+
+    @property
+    def index(self) -> int:
+        """Numeric index of the %patch macro."""
+        if self.options.P is not None:
+            return self.options.P
+        tokens = re.split(r"(\d+)", self._name, maxsplit=1)
+        if len(tokens) > 1:
+            return int(tokens[1])
+        return 0
+
+
+class AutosetupMacro(PrepMacro):
+    """Class that represents an %autosetup macro."""
+
+    def _setup_parser(self) -> None:
+        """Configures internal `ArgumentParser` for the options of the macro."""
+        self._parser.add_argument("-n", default="%{name}-%{version}")
+        self._parser.add_argument("-v", action="store_true")
+        self._parser.add_argument("-c", action="store_true")
+        self._parser.add_argument("-D", action="store_true")
+        self._parser.add_argument("-T", action="store_true")
+        self._parser.add_argument("-b", type=int)
+        self._parser.add_argument("-a", type=int)
+        self._parser.add_argument("-N", action="store_true")
+        self._parser.add_argument("-S", default="patch")
+        self._parser.add_argument("-p", type=int)
+
+
+class AutopatchMacro(PrepMacro):
+    """Class that represents an %autopatch macro."""
+
+    def _setup_parser(self) -> None:
+        """Configures internal `ArgumentParser` for the options of the macro."""
+        self._parser.add_argument("-v", action="store_true")
+        self._parser.add_argument("-p", type=int)
+        self._parser.add_argument("-m", type=int)
+        self._parser.add_argument("-M", type=int)
+        self._parser.add_argument("indices", type=int, nargs="*")
+
+
+class PrepMacros(collections.abc.Sequence):
+    """Class that represents a sequence of all %prep macros."""
+
+    def __init__(self, section: Section) -> None:
+        """
+        Constructs a `PrepMacros` object.
+
+        Args:
+            section: %prep section.
+
+        Returns:
+            Constructed instance of `PrepMacros` class.
+        """
+        self._section = section
+
+    def __repr__(self) -> str:
+        section = repr(self._section)
+        return f"PrepMacros({section})"
+
+    def __len__(self) -> int:
+        return len(self._get_items())
+
+    def _get_items(self) -> List[PrepMacro]:
+        """
+        Gets all supported %prep macros.
+
+        Returns:
+            List of instances of subclasses of PrepMacro.
+        """
+        comment_regex = re.compile(r"^\s*#.*$")
+        macro_regex = re.compile(
+            r"(?P<c>%{!?\?\w+:)?.*?"
```

Comment:
I'm a bit lost in this regex. The second group (the macro itself) and the last (options) are clear but what about the first group?

Reply:
That's the start of a conditionalized macro expansion, it needs to be there to be able to match the ending bracket to ensure it's not included in the option string.

Real world example: `%{?with_system_nss:%patch30 -p3 -b .nss_pkcs11_v3}`

I'll add a comment.